t/14-grutasks.t

```diff
@@ -500,19 +502,48 @@ subtest 'Gru manual task' => sub {
       'minion job has the right error message';
 };
 
-SKIP: {
-    skip 'no network available', 1 if $ENV{OBS_RUN};
-    subtest 'download assets with correct permissions' => sub {
-        # need to whitelist github
-        $t->app->config->{global}->{download_domains} = 'github.com';
-
-        my $assetsource = 'https://github.com/os-autoinst/os-autoinst/blob/master/t/data/Core-7.2.iso';
-        my $assetpath   = 't/data/openqa/share/factory/iso/Core-7.2.iso';
-        run_gru('download_asset' => [$assetsource, $assetpath, 0]);
-        ok(-f $assetpath, 'asset downloaded');
-        is(S_IMODE((stat($assetpath))[2]), 0644, 'asset downloaded with correct permissions');
-    };
-}
+# prevent writing to a log file to enable use of combined_like in the following tests
+$t->app->log(Mojo::Log->new(level => 'debug'));
+
+subtest 'download assets with correct permissions' => sub {
+    my $assetsource = 'https://github.com/os-autoinst/os-autoinst/blob/master/t/data/Core-7.2.iso';
+    my $assetpath   = 't/data/openqa/share/factory/iso/Core-7.2.iso';
+
+    # be sure the asset does not exist from a previous test run
+    unlink($assetpath);
+
+    combined_like(
+        sub {
+            run_gru('download_asset' => [$assetsource, $assetpath, 0]);
+        },
+        qr/host github\.com .* is not on the whitelist \(which is empty\)/,
+        'download refused if whitelist empty',
+    );
+
+    $t->app->config->{global}->{download_domains} = 'gitlab.com';
+    combined_like(
+        sub {
+            run_gru('download_asset' => [$assetsource, $assetpath, 0]);
+        },
+        qr/host github\.com .* is not on the whitelist/,
+        'download refused if host not on whitelist',
+    );
+
+    plan skip_all => 'no network available' if $ENV{OBS_RUN};
+
+    $t->app->config->{global}->{download_domains} .= ' github.com';
+    combined_like(
+        sub {
+            run_gru('download_asset' => [$assetsource . '.foo', $assetpath, 0]);
+        },
+        qr/404 response\:/,
+        'error code logged',
+    );
+
+    run_gru('download_asset' => [$assetsource, $assetpath, 0]);
+    ok(-f $assetpath, 'asset downloaded');
+    is(S_IMODE((stat($assetpath))[2]), 0644, 'asset downloaded with correct permissions');
```

Comment:
I would like it more if we download from a local server. I guess we have enough around within the openQA context running already :) Or, to not download from the actual openQA server to test, you could setup a trivial mojo webserver offering a local file. Something smaller than the (small) ISO would also be good.

I don't want openQA tests to cost me a fortune when running them over a mobile data connection from train ;)